wlroots/wlr_types/layer_shell_v1.py

```diff
@@ -0,0 +1,213 @@
+
+    @output.setter
+    def output(self, output: Output):
+        self._ptr.output = output._ptr
+
+    @property
+    def added(self) -> bool:
+        return self._ptr.added
+
+    @property
+    def configured(self) -> bool:
+        return self._ptr.configured
+
+    @property
+    def mapped(self) -> bool:
+        return self._ptr.mapped
+
+    @property
+    def closed(self) -> bool:
+        return self._ptr.closed
+
+    @property
+    def client_pending(self) -> LayerSurfaceV1State:
+        state_ptr = self._ptr.client_pending
+        _weakkeydict[state_ptr] = self._ptr
+        return LayerSurfaceV1State(state_ptr)
+
+    @property
+    def server_pending(self) -> LayerSurfaceV1State:
+        state_ptr = self._ptr.server_pending
+        _weakkeydict[state_ptr] = self._ptr
+        return LayerSurfaceV1State(state_ptr)
+
+    @property
+    def current(self) -> LayerSurfaceV1State:
+        state_ptr = self._ptr.current
+        _weakkeydict[state_ptr] = self._ptr
+        return LayerSurfaceV1State(state_ptr)
+
+    def configure(self, width: int, height: int) -> None:
+        """
+        Notifies the layer surface to configure itself with this width/height. The
+        layer_surface will signal its map event when the surface is ready to assume
+        this size.
+        """
+        lib.wlr_layer_surface_v1_configure(self._ptr, width, height)
+
+    def close(self) -> None:
+        lib.wlr_layer_surface_v1_close(self._ptr)
+
+    @classmethod
+    def from_wlr_surface(cls, surface: Surface):
+        surface_ptr = lib.wlr_layer_surface_v1_from_wlr_surface(surface._ptr)
+        _weakkeydict[surface_ptr] = surface._ptr
+        return LayerSurfaceV1(surface_ptr)
+
+    def for_each_surface(
+        self, iterator: SurfaceCallback[T], data: Optional[T] = None
+    ) -> None:
+        """
+        Calls the iterator function for each sub-surface and popup of this surface
+        """
+        py_handle = (iterator, data)
+        handle = ffi.new_handle(py_handle)
+        lib.wlr_layer_surface_v1_for_each_surface(
+            self._ptr, lib.surface_iterator_callback, handle
+        )
+
+    def surface_at(
+        self, sx: float, sy: float
+    ) -> Tuple[Optional[Surface], float, float]:
+        """
+        Find a surface within this layer-surface tree at the given surface-local
+        coordinates. Returns the surface and coordinates in the leaf surface
+        coordinate system or None if no surface is found at that location.
+        """
+        sub_x_data = ffi.new("double*")
+        sub_y_data = ffi.new("double*")
+        surface_ptr = lib.wlr_layer_surface_v1_surface_at(
+            self._ptr, sx, sy, sub_x_data, sub_y_data
+        )
+        if surface_ptr == ffi.NULL:
+            return None, 0.0, 0.0
+
+        return Surface(surface_ptr), sub_x_data[0], sub_y_data[0]
+
+
+def wlr_surface_is_layer_surface(surface: Surface) -> bool:
+    return lib.wlr_surface_is_layer_surface(surface._ptr)
```

Comment:
I'd added something like this as a property of the `Surface` to see if it an XDG surface. I'm not sure if it makes sense to add all of these different handlers to `Surface` for the checks, or if they make more sense to be defined more like this.

Reply:
I like the idea of them being kept separate like the original code, but at the same it definitely makes a more unified and nice OO interface if it was a method. That gets my vote.